source/distributed/shared_tria.cc

```diff
@@ -134,6 +135,20 @@ namespace parallel
       this->update_number_cache ();
     }
 
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::
+    copy_triangulation (const dealii::Triangulation<dim, spacedim> &old_tria)
+    {
+      Assert ((dynamic_cast<const dealii::parallel::distributed::Triangulation<dim,spacedim> *>(&old_tria) == NULL),
+              ExcMessage("Cannot use this function on parallel::distributed::Triangulation."));
```

Comment:
> I think it is enough.

No, the point of assertions is to be defensive. I have learned in the past, that it is much better to trigger an assert that might be unnecessary instead of weird/broken behavior. That way you can think about it, write a test, confirm that it works, etc.. Note that this is most likely to happen to somebody who did not write this code.